### PR TITLE
Implicit nested transaction from transactional methods inherit transactional context parent type.

### PR DESCRIPTION
**Pull request by rogermichoud**

### CODE REVIEW

**Comment by rogermichoud** [comment]
#677

File: runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java [hunk 1]
```diff
@@ -302,7 +302,13 @@ public UUID getStreamID() {
         int retries = 1;
         while (true) {
             try {
-                rt.getObjectsView().TXBegin();
+                /* If it is a nested transaction, inherit type of parent */
+                if (TransactionalContext.getCurrentContext() != null) {
```
Comment (Maithem) [inline runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java]: I think this should be fixed in a higher layer (i.e. ObjectsView class).

Consider something like:


@@ -105,8 +106,14 @@ public class ObjectsView extends AbstractView {
      * to other threads or clients until TXEnd is called.
      */
     public void TXBegin() {
+        TransactionType type = TransactionType.OPTIMISTIC;
+
+        if(TransactionalContext.isInTransaction()) {
+            type = TransactionalContext.getCurrentContext().getBuilder().type;
+        }
+
         TXBuild()
-                .setType(TransactionType.OPTIMISTIC)
+                .setType(type)
                 .begin();
     }

**Comment by Maithem** [comment]
Also, pkease fix git commit message 
https://chris.beams.io/posts/git-commit/

**Comment by Maithem** [comment]
Hmm, I guess its not only a matter of preference, but it was also wrong. Your last patch was failing this test

    @Test
    public void testNested() {
        CorfuRuntime r = getDefaultRuntime();
        Map<String, String> smrMap = r.getObjectsView().build()
                .setStreamName("map a")
                .setTypeToken(new TypeToken<SMRMap<String, String>>() {})
                .open();
        r.getObjectsView().TXBuild()
                .setType(TransactionType.SNAPSHOT)
                .begin();

        r.getObjectsView().TXBegin();

        //smrMap.forEach((k,v)->System.out.println("k : " + k + " v : " + v));

        r.getObjectsView().TXEnd();
    }

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **3cb72e2**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #678 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/kZmyQnS0Q40I3oM9cQSyrtaFkbFqrxNy)

**Comment by Maithem** [comment]
addressed by https://github.com/CorfuDB/CorfuDB/pull/679

**Pull request closed without merging**